torch/csrc/jit/codegen/cuda/runtime/tensor_core.cu

```diff
@@ -0,0 +1,410 @@
+
+// Arch guard, Volta + only
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 700))
+
+namespace mma_instruction {
+
+// MMA instruction wrappers (sm_70+):
+// The instruction wrappers below are quarter-warp macros, which currently
+// nvfuser
+//  doesn't explicitly model. So they are currently only meant to be
+//  used as building blocks in warp level mma macros
+
+//  8x8x4 mma instruction, per quarter warp (8 threads), fp16 accumulate
+//  per thread register:
+//   A[4] x B[4] -> C[8]
+__device__ inline void mmaM8n8k4(__half* C, __half* A, __half* B) {
+  unsigned const* _A = reinterpret_cast<unsigned const*>(A);
+  unsigned const* _B = reinterpret_cast<unsigned const*>(B);
+  unsigned* _C = reinterpret_cast<unsigned*>(C);
+
+  asm("mma.sync.aligned.m8n8k4.row.row.f16.f16.f16.f16 {%0,%1,%2,%3}, {%4,%5}, {%6,%7}, {%8,%9,%10,%11};\n"
+      : "=r"(_C[0]), "=r"(_C[1]), "=r"(_C[2]), "=r"(_C[3])
+      : "r"(_A[0]),
+        "r"(_A[1]),
+        "r"(_B[0]),
+        "r"(_B[1]),
+        "r"(_C[0]),
+        "r"(_C[1]),
+        "r"(_C[2]),
+        "r"(_C[3]));
+}
+
+//  8x8x4 mma instruction, per quarter warp (8 threads), fp32 accumulate
+//  per thread register:
+//   A[4] x B[4] -> C[8]
+__device__ inline void mmaM8n8k4(float* C, __half* A, __half* B) {
+  unsigned const* _A = reinterpret_cast<unsigned const*>(A);
+  unsigned const* _B = reinterpret_cast<unsigned const*>(B);
+
+  asm("mma.sync.aligned.m8n8k4.row.row.f32.f16.f16.f32 {%0,%1,%2,%3,%4,%5,%6,%7}, {%8,%9}, {%10,%11}, {%12,%13,%14,%15,%16,%17,%18,%19};\n"
+      : "=f"(C[0]),
+        "=f"(C[1]),
+        "=f"(C[2]),
+        "=f"(C[3]),
+        "=f"(C[4]),
+        "=f"(C[5]),
+        "=f"(C[6]),
+        "=f"(C[7])
+      : "r"(_A[0]),
+        "r"(_A[1]),
+        "r"(_B[0]),
+        "r"(_B[1]),
+        "f"(C[0]),
+        "f"(C[1]),
+        "f"(C[2]),
+        "f"(C[3]),
+        "f"(C[4]),
+        "f"(C[5]),
+        "f"(C[6]),
+        "f"(C[7]));
+}
+
+} // namespace mma_instruction
+
+// Swizzling utilities, currently only mma using it, could consider
+//  moving to a common space if needed.
+namespace mem_swizzle {
+
+// Basic coordinate unit for 2D swizzling
+struct MatrixCoordinate {
+  nvfuser_index_t x;
+  nvfuser_index_t y;
+
+  __device__ MatrixCoordinate(nvfuser_index_t x_, nvfuser_index_t y_)
+      : x(x_), y(y_) {}
+  __device__ MatrixCoordinate(const MatrixCoordinate& other)
+      : x(other.x), y(other.y) {}
+
+  // Helper function to index shared mem,
+  //   if we decided to wrap smem pointers in a class should probably move this
+  //   there.
+  __device__ inline nvfuser_index_t linearize(const MatrixCoordinate& stride) {
```

Comment:
It would be good to figure out an interface where we can return the 2D swizzled indices and let the codegen handle the linearization. This is useful if the swizzle coordinates are not the innermost-contiguous coordinates.